Move host and tracer port to the agent settings group

marcotc opened the pull request:
This PR moves the `c.tracer.hostname` and `c.tracer.port` options to a more meaningful namespace.

`c.tracer.hostname` is not tracer specific, thus moved to `c.agent.host`.
`c.tracer.port` refers to the APM agent port, thus moved to `c.agent.port`.

**Breaking changes**

- Changed: Configuration setting `c.tracer.hostname` moved to `c.agent.host`.
- Changed: Configuration setting `c.tracer.port` moved to `c.agent.port`.

ivoanjo commented on lib/ddtrace/configuration/settings.rb:
```diff
@@ -44,6 +44,39 @@ def initialize(*_)
+        settings :tracer do
+          # Trace agent port.
+          # @configure_with {Datadog::Tracing}
+          # @default `DD_AGENT_HOST` environment variable, otherwise `8126`
+          # @return [String,nil]
+          option :port
```
```suggestion
          # @default `DD_TRACE_AGENT_PORT` environment variable, otherwise `8126`
          # @return [String,nil]
          option :port
```

ivoanjo commented on lib/ddtrace/configuration/settings.rb:
```diff
@@ -44,6 +44,39 @@ def initialize(*_)
+          option :port
+        end
+
+        # TODO: add declarative statsd configuration. Currently on environment variable configuration is possible.
```
```suggestion
        # TODO: add declarative statsd configuration. Currently only usable via an environment variable.
```

ivoanjo reviewed (commented):
I'm a bit ambivalent here: I would prefer to add only a global `agent.host` and `agent.port` configuration for 1.0.0, and then later adding a tracer-specific port.

This is because the current approach requires:
1- Adding a settings group that is both global and product-specific
2- Adding one more inconsistency. Forget that you are a Datadog developer. Look at my tracing library -- LibFoo. Tell me what you think:

```ruby
LibFoo::Tracing.configure do |c| # Ok I'm inside the tracing-specific configuration
  c.tracer.enabled = true        # Wait, why tracer inside tracing? Shouldn't this be c.enabled?
  c.agent.tracer.port = 8126     # Wait, what is the agent settings? Am I not inside tracing? How does c.agent.port differ from c.agent.tracer.port?
  c.sampling.default_rate = 1.0  # Ok this does not say tracer. Is that relevant? Should I be looking at a c.tracer.sampling.default_rate or a c.sampling.tracer.default_rate instead?
```

On the other hand:

3- I realize we've named the env var in a bunch of tracing libraries `DD_TRACE_AGENT_PORT` (although it looks like python has a `DD_AGENT_PORT`!)
4- If the choice is between this PR and what we had before, I do think this one is slightly better.

So to clarify, here's what I suggest we do for this PR. My suggestion is not blocking -- I can live with the PR as-is, see my notes above.

```ruby
Datadog.configure do |c| # this config is global, for all products
  c.agent.host = 'foo'
  c.agent.port = '12345'
end
```

And here's how I envision we could evolve it later:

```ruby
Datadog.configure do |c| # this config is global, for all products
  c.agent.host = 'foo'
  c.agent.port = '12345'
end

Datadog::Tracing.configure do |c|
  c.agent.port = '456' # overrides the global config just for tracing; other products still use 12345
end
```

Thoughts?

marcotc commented:
@ivoanjo one issue is that there are (currently) two distinct agent ports: tracer port and statsd port. So there's no canonical "agent port". That makes naming super tricky.

delner commented:
> And here's how I envision we could evolve it later

@ivoanjo I agree with this as a final destination (at least as far as I can see.) But the point remains that we can't implement this without some much more serious configuration overhaul, which is currently out of reach.

While there would be a smoother transition from `agent.port` to a `Tracing.configure { agent.port }`, the issue is `agent.port` is ambiguous; there is no "global" agent port, ports are specific to features and can be configured separately. So what port would `agent.port` be?

I think as its written is more clear in its function in the interim; thus I'd prefer that. But I prefer your vision long term if we can implement that later.

ivoanjo commented:
> @ivoanjo one issue is that there are (currently) two distinct agent ports: tracer port and statsd port. So there's no canonical "agent port". That makes naming super tricky.

> The issue is agent.port is ambiguous in the interim; there is no "global" agent port, ports are specific to features and can be configured separately. So what port would agent.port be?

Ahaha very interesting! I actually have a completely different view on it, perhaps from having come in more recently, and being on the edges of apm teams :)

I've always thought as it being a difference between protocols: 

1. One port for talking to the agent using the statsd/dogstatsd protocol (statsd only)
2. One port for talking to the agent via http (tracing, profiling, others...(?) aka everything else)

The same change kinda maps to the unix domain socket support, when used: there's an `apm.socket` (HTTP over unix domain socket) and an `dsd.socket` (statsd over unix domain socket).

Which is why I've been annoying y'all about having a global config for port. (And I nominate the http port to be the global one since the other is only for the statsd protocol).

So even considering statsd, my full vision would be:

```ruby
Datadog.configure do |c|
  c.agent.port = '1234' # the port for http
end

Datadog::Statsd.configure do |c|
  c.statsd.port = '5678'
end

Datadog::Tracing.configure do |c|
  c.agent.port = '91011' # override the http port on a product-by-product basis
end
```

> @ivoanjo I agree with this as a final destination (at least as far as I can see.) But the point remains that we can't implement this without some much more serious configuration overhaul, which is currently out of reach.

Actually, I think we're quite close here. The [`AgentSettingsResolver`](https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/configuration/agent_settings_resolver.rb#L20) is being used by both the tracer and the profiler.

In this PR, we moved the `hostname` from inside tracer to the global configuration. If we moved the `port` as well, and for now considered both of them to be global configs only, we could be at

```ruby
Datadog.configure do |c| # this config is global, for all products
  c.agent.host = 'foo'
  c.agent.port = '12345'
end
```
today. I estimate that this would be a diff of around 5-10 lines on top of this current PR, and everything would continue to work.

---

Now, am I looking at this all wrong? Is the "http" port/"apm.socket" not really the everything-else port but just a really-really-badly-named "traces port" and "traces.socket"?

If so, then yeah, what I said above makes no sense and this PR should go as-is :)

ivoanjo commented on lib/datadog/core/configuration/settings.rb:
```diff
@@ -47,6 +47,39 @@ def initialize(*_)
+          # @return [String,nil]
+          option :host
+
+          # Configuration for APM agent access.
```
Same note as above re: calling it "APM Agent"

ivoanjo commented on lib/datadog/core/configuration/settings.rb:
```diff
@@ -47,6 +47,39 @@ def initialize(*_)
+          option :host
+
+          # Configuration for APM agent access.
+          # @configure_with {Datadog::Tracing}
+          # @public_api
+          settings :apm do
+            # Agent APM TCP port.
+            # @configure_with {Datadog::Tracing}
```
Both `@configure_with` are outdated

ivoanjo reviewed (commented):
Thanks for looking into this. I'll still try to push back on making it `c.agent.apm.port`, but if you feel strongly about it, I can live with it :)

delner commented:
Not to push the goal line on this here, but I don't think we're supposed to add team-like names (`apm`) into code like this. However, if that's what the agent is named ("APM Agent") in our public documentation, then it seems reasonable.

ivoanjo commented:
> However, if that's what the agent is named ("APM Agent") in our public documentation, then it seems reasonable.

We may have checked different places, but I didn't see it being called APM Agent in the public docs (which was part of my reasoning to not use that name)

marcotc commented on docs/GettingStarted.md:
```diff
@@ -1885,15 +1885,15 @@ To change the default behavior of the Datadog tracer, you can provide custom opt
 # Global settings are set here:
 Datadog.configure do |c|
+  c.agent.host = 'custom-agent-host'
+  c.agent.apm.port = 8126
```
Which agent port would this be referring to (from: https://docs.datadoghq.com/agent/guide/network/?tab=agentv6v7#open-ports):
![Screen Shot 2022-01-28 at 11 21 04 AM](https://user-images.githubusercontent.com/583503/151608450-55c2469d-7c9e-4837-886b-20795500e4f8.png)

`ddtrace` currently uses two of these: `8125` UDP for metrics and `8126` TCP for traces and profilers.

marcotc replied:
I changed the setting to `c.agent.port`, follow Node's example.

marcotc commented:
@delner @ivoanjo I got the APM agent nomenclature after talking to the agent team.

Additionally, when users configure their agents today, the configuration in question here is called `apm_config: receiver_port: 8126`: https://github.com/DataDog/datadog-agent/blob/61b2fe449e42b5f8e8e8d4b7646894477a9dd447/pkg/config/config_template.yaml#L976-L997

Similarly in a Kubernetes setup, the option is called `apm: port: 8126`: https://github.com/DataDog/helm-charts/blob/24a79128ffb67a5fd48299ad153ed1f879db4167/charts/datadog/values.yaml#L265-L281

This PR is trying to keep our nomenclature consistent across Datadog products.

If the nomenclature changes, we'll adapt as well.

marcotc commented:
The alternatives seem worse to me:

1. Simply call it port: There are many other agent ports, including ports used by `ddtrace` itself for metrics.
2. Call it tracer port: Matches the historical trace agent nomenclature, but doesn't match current public agent configuration language.

marcotc commented:
Node has `port` for the tracing port, and `dogstatsd.port` for metrics port: https://docs.datadoghq.com/tracing/setup_overview/setup/nodejs/?tab=containers#instrumentation
That's also an option.

marcotc commented:
Ok, simply port seems fine then, and keep statsd as a separate concern (which is won't be addressed today, given it's not configurable at this point).

marcotc commented on docs/GettingStarted.md:
```diff
@@ -1912,10 +1912,10 @@ end
  - `instance`: set to a custom `Datadog::Tracer` instance. If provided, other trace settings are ignored (you must configure it manually.)
  - `partial_flush.enabled`: set to `true` to enable partial trace flushing (for long running traces.) Disabled by default. *Experimental.*
- - `port`: set the port the trace agent is listening on.
+ - `agent.apm.port`: set the TCP port the APM agent is listening on.
```
I clarified this a bit, let me know if this works better now.

marcotc commented:
I've renamed `agent.apm.port` to `agent.port`.

The language around "APM port in agent" is consistent with our Agent product documentation, so I kept it like that. I removed mentions to a "apm agent", instead having "apm port in agent" or equivalent language.

ivoanjo commented on spec/ddtrace/integration_spec.rb:
```diff
@@ -493,8 +493,8 @@ def agent_receives_span_step3(previous_success)
     subject(:configure) do
       Datadog::Tracing.configure do |c|
-        c.tracer.hostname = hostname
-        c.tracer.port = port
+        c.agent.host = hostname
+        c.agent.port = port
```
This seems like it will/should fail, since it's configuring these from inside `Datadog::Tracing.configure`.

marcotc replied:
This 100% explodes, thank you!

ivoanjo reviewed (approved):
👍 Thanks for digging deep on this.

To be honest, your research even got me convinced that "apm port" is quite reasonable as well: based on your research I'm happy with either "agent port" or "apm port".
